File: 3-longest-substring-without-repeating-characters/longest-substring-without-repeating-characters.cpp
class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        int l=0, r=0, maxi = 0;
        int n = s.size();
        vector<int> hash(256, -1);
        while(r < n){
            if(hash[s[r]]>=l){
                l=hash[s[r]]+1;
            }
            hash[s[r]] = r;
            maxi = max(maxi, r-l+1);
            r++;
        }
        return maxi;
    }
};

File: 3-longest-substring-without-repeating-characters/longest-substring-without-repeating-characters.java
class Solution {
    public int lengthOfLongestSubstring(String s) {
        int n = s.length();
        if(n == 0) return 0;
        int i=0, j=0, maxi = 1;
        Map<Character, Integer> map = new HashMap<>(); 
        while(j<n) {
            if(map.get(s.charAt(j)) != null) {
                if(map.get(s.charAt(j)) >= i) {
                    i = map.get(s.charAt(j)) + 1;
                }
            }
            map.put(s.charAt(j), j);
            maxi = Math.max(maxi, j-i+1);
            j++;
        }
        return maxi;
    }
}